client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx

```diff
@@ -0,0 +1,226 @@
+
+		const payload = {
+			...this.state.form,
+			role: 'main',
+		};
+
+		const errors = Object.assign(
+			! payload.host && { host: translate( 'Please enter a valid server address.' ) },
+			! payload.port && { port: translate( 'Please enter a valid server port.' ) },
+			isNaN( payload.port ) && { port: translate( 'Port number must be numeric.' ) },
+			! payload.user && { user: translate( 'Please enter your server username.' ) },
+			! payload.pass && { pass: translate( 'Please enter your server password.' ) },
+			! payload.abspath && { abspath: translate( 'Please enter a valid upload path.' ) }
+		);
+
+		return isEmpty( errors )
+			? updateCredentials( siteId, payload )
+			: this.setState( { formErrors: errors } );
+	};
+
+	togglePrivateKeyField = () =>
+		this.setState( { showPrivateKeyField: ! this.state.showPrivateKeyField } );
+
+	renderInputField = ( name, title, placeholder ) => {
+		const { formIsSubmitting, onCancel, translate } = this.props;
+		const { formErrors } = this.state;
+
+		return (
+			<FormLabel>
+				<div>{ title }</div>
+				<FormTextInput
+					name={ name }
+					placeholder={ placeholder }
+					value={ get( this.state.form, name, '' ) }
+					onChange={ this.handleFieldChange }
+					disabled={ formIsSubmitting }
+					isError={ !! formErrors[ name ] }
+				/>
+				{ formErrors[ name ] && (
+					<FormInputValidation isError={ true } text={ formErrors[ name ] } />
+				) }
+			</FormLabel>
+		);
+	};
+
+	render() {
+		const { formIsSubmitting, onCancel, translate } = this.props;
+
+		const { showPrivateKeyField, formErrors } = this.state;
+
+		return (
+			<FormFieldset className="credentials-form">
+				<table>
+					<tbody>
+						<tr>
+							<td colSpan="2" className="credentials-form__protocol-field">
+								<FormLabel>
+									<div>{ translate( 'Credential Type' ) }</div>
+									<FormSelect
+										name="protocol"
+										value={ get( this.state.form, 'protocol', 'ssh' ) }
+										onChange={ this.handleFieldChange }
+										disabled={ formIsSubmitting }
+									>
+										<option value="ssh">{ translate( 'SSH' ) }</option>
+										<option value="sftp">{ translate( 'SFTP' ) }</option>
+										<option value="ftp">{ translate( 'FTP' ) }</option>
+									</FormSelect>
+								</FormLabel>
+							</td>
+						</tr>
+						<tr>
+							<td className="credentials-form__host-field">
+								{ this.renderInputField(
+									'host',
+									translate( 'Server Address' ),
+									translate( 'yoursite.com' )
+								) }
+							</td>
+							<td className="credentials-form__port-field">
+								{ this.renderInputField( 'port', translate( 'Port Number' ), translate( '22' ) ) }
+							</td>
+						</tr>
+						<tr>
+							<td colSpan="2" className="credentials-form__user-field">
+								{ this.renderInputField(
+									'user',
+									translate( 'Username' ),
+									translate( 'username' )
+								) }
+							</td>
```

Comment:
If we only get two fields instead of all the fields out of it, we don't have to extract the JSX from here. When I first raised the idea I thought it might be possible to basically cover all the fields, `<tr>`, `<td>`, and others in such a way that it cuts out all of the manual bits.

If those bits remain _and_ we have a new function it may not be worth it, unless we can go all the way.